File: LocationManager/LocationService.swift
//
//  LocationService.swift
//  SwiftiDate
//
//  Created by 游哲維 on 2025/2/16.
//

import Foundation
import CoreLocation
import UIKit

class LocationService: NSObject, CLLocationManagerDelegate {
    
    private var locationManager: CLLocationManager
    private let geocoder = CLGeocoder()
    
    // 這裡可用自訂 closure 或 Notification 來回傳更新的地址或緯經度
    var onLocationUpdate: ((CLLocation) -> Void)?
    var onPlacemarkUpdate: ((CLPlacemark) -> Void)?
    
    override init() {
        self.locationManager = CLLocationManager()
        super.init()
        self.locationManager.delegate = self
    }
    
    func start() {
        locationManager.requestWhenInUseAuthorization()
        locationManager.startUpdatingLocation()
    }
    
    func stop() {
        locationManager.stopUpdatingLocation()
    }
    
    // MARK: - CLLocationManagerDelegate
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        
        // 可以透過 closure 或 Notification 傳遞給外部
        onLocationUpdate?(location)
        
        // 埋點：成功取得經緯度
        AnalyticsManager.shared.trackEvent("location_update", parameters: [
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude
        ])
        
        // 反向地理編碼
        geocoder.reverseGeocodeLocation(location) { [weak self] placemarks, error in
            if let error = error {
                print("Reverse geocode failed: \(error.localizedDescription)")
                // 可以在失敗時也做事件紀錄
                AnalyticsManager.shared.trackEvent("reverse_geocode_failed", parameters: [
                    "error": error.localizedDescription
                ])
                return
            }
            if let placemark = placemarks?.first {
                self?.onPlacemarkUpdate?(placemark)
                
                AnalyticsManager.shared.trackEvent("reverse_geocode_success", parameters: [
                    "country": placemark.country ?? "unknown",
                    "locality": placemark.locality ?? "unknown",
                    "coordinate": "\(location.coordinate.latitude),\(location.coordinate.longitude)"
                ])
            }
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("Failed to get location: \(error.localizedDescription)")
    }
}

File: flutter/swiftidate/lib/services/location_service.dart
// lib/services/location_service.dart
import 'package:geolocator/geolocator.dart';
import 'package:geocoding/geocoding.dart';
import '../analytics_manager.dart';

class LocationService {
  // Callback 供外部註冊：傳回最新位置
  Function(Position)? onLocationUpdate;
  // Callback 供外部註冊：傳回反向地理編碼的結果
  Function(Placemark)? onPlacemarkUpdate;

  // 開始取得位置更新並進行反向地理編碼
  Future<void> start() async {
    // 請求權限
    await Geolocator.requestPermission();
    // 開始位置更新
    Geolocator.getPositionStream(
      locationSettings: const LocationSettings(
        accuracy: LocationAccuracy.best,
      ),
    ).listen((position) {
      if (onLocationUpdate != null) {
        onLocationUpdate!(position);
      }
      AnalyticsManager.shared.trackEvent("location_update", parameters: {
        "latitude": position.latitude,
        "longitude": position.longitude,
      });
      _reverseGeocode(position);
    });
  }

  Future<void> _reverseGeocode(Position position) async {
    try {
      List<Placemark> placemarks = await placemarkFromCoordinates(
        position.latitude,
        position.longitude,
      );
      if (placemarks.isNotEmpty) {
        final placemark = placemarks.first;
        if (onPlacemarkUpdate != null) {
          onPlacemarkUpdate!(placemark);
        }
        AnalyticsManager.shared.trackEvent("reverse_geocode_success", parameters: {
          "country": placemark.country ?? "unknown",
          "locality": placemark.locality ?? "unknown",
          "coordinate": "${position.latitude},${position.longitude}",
        });
      }
    } catch (error) {
      print("Reverse geocode failed: $error");
      AnalyticsManager.shared.trackEvent("reverse_geocode_failed", parameters: {
        "error": error.toString(),
      });
    }
  }

  // 若需要停止更新，可加入取消訂閱的邏輯
}
